### ISSUE TITLE
`arguments-out-of-order` false positive? (W1114)

### ISSUE THREAD

**Issue by turbotimon** [PyCQA/pylint#8001]
### Bug description

This is the code with arguments in perfect oder and same names:

```python
# first line is line number 36
def create_link(osm: OsmObject, wd: WdItem, key: str) -> Link:
    if isinstance(osm, OsmNode):
        return NodeLink(osm, wd, key) # line 38
    if isinstance(osm, OsmWay):
        return WayLink(osm, wd, key) # line 40
    if isinstance(osm, OsmRelation):
        return RelationLink(osm, wd, key) # line 42
    if isinstance(osm, OsmArea):
        return AreaLink(osm, wd, key) # line 44
    raise TypeError()
```

Class Definition is:
```
@dataclass
class Link(ABC):
    osm: OsmObject
    wd: WdItem
    key: str

@dataclass
class NodeLink(Link):
    osm: OsmNode

# ... same for the others
```

### Configuration

```ini
[tool.pylint.'MESSAGES CONTROL']
disable="R0914, R0801, C0116, C0115, C0114, C0301, logging-fstring-interpolation, no-method-argument, too-few-public-methods, too-many-return-statements, invalid-name, fixme, too-many-arguments, consider-using-with, too-many-instance-attributes, missing-timout"

[tool.pylint.MASTER]
extension-pkg-whitelist="osmium"
```

### Command used

```shell
pylint <name>
```

### Pylint output

```shell
************* Module <name>.checker_issues
<name>/checker_issues.py:38:15: W1114: Positional arguments appear to be out of order (arguments-out-of-order)
<name>/checker_issues.py:40:15: W1114: Positional arguments appear to be out of order (arguments-out-of-order)
<name>/checker_issues.py:42:15: W1114: Positional arguments appear to be out of order (arguments-out-of-order)
<name>/checker_issues.py:44:15: W1114: Positional arguments appear to be out of order (arguments-out-of-order)
```

### Expected behavior

not issues

### Pylint version

```shell
pylint 2.15.9
astroid 2.12.13
Python 3.10.2 (main, Dec 29 2022, 11:10:35) [GCC 9.4.0
```

### OS / Environment

wsl:ubuntu

**Comment by clavedeluna** [comment]
Partial reproducer
```
from dataclasses import dataclass
from abc import ABC

class OsmObject:
    pass

class WdItem:
    pass

class OsmNode:
    pass

class OsmWay:
    pass

class WayLink:
    pass

@dataclass
class Link(ABC):
    osm: OsmObject
    wd: WdItem
    key: str

@dataclass
class NodeLink(Link):
    osm: OsmNode

def create_link(osm: OsmObject, wd: WdItem, key: str) -> Link:
    if isinstance(osm, OsmNode):
        return NodeLink(osm, wd, key) # line 38
    if isinstance(osm, OsmWay):
        return WayLink(osm, wd, key) # line 40
    # if isinstance(osm, OsmRelation):
    #     return RelationLink(osm, wd, key) # line 42
    # if isinstance(osm, OsmArea):
    #     return AreaLink(osm, wd, key) # line 44
    raise TypeError()
```

A removal of @dataclass appears to make this go away, so it may be related to dataclass.
